### ISSUE TITLE
Indicate field is nullable in swagger/openapi

### ISSUE THREAD

**Issue by jurasofish** [tiangolo/fastapi#2924]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI
from pydantic import BaseModel, Field
from typing import Optional

app = FastAPI()


class MyResponseAmbiguous(BaseModel):
    a: Optional[int]

@app.get("/ambiguous", response_model=MyResponseAmbiguous)
def ambiguous():
    # swagger shows ``a`` as type int, even though it can be None
    return MyResponseAmbiguous(a=1)



class MyResponseNullable(BaseModel):
    a: Optional[int] = Field(..., nullable=True)

@app.get("/nullable", response_model=MyResponseNullable)
def nullable():
    # swagger shows ``a`` as type int, and also says it can be nullable.
    return MyResponseNullable(a=1)

```

### Description

When generating an openapi schema, FastAPI does not mark ``Optional`` items as nullable - as in the ``MyResponseAmbiguous`` model above.
To mark an item as nullable in openapi you need to manually pass ``nullable=True`` to the pydantic Field function - as in the ``MyResponseNullable`` model above.
How can this be done automatically?

Additionally, one of the models has ``a`` marked as required (red asterisk) and the other does not - is this the intended behaviour? I think it should be marked as required in both models.

<img width="401" alt="Screen Shot 2021-03-10 at 4 45 25 pm" src="https://user-images.githubusercontent.com/33386122/110582238-05721a80-81c0-11eb-9afb-6b9e62265a2e.png">

### Environment

* OS: Macos
* FastAPI Version: 0.63.0
* Python version: 3.7

**Comment by Mause** [comment]
This is a duplicate of https://github.com/tiangolo/fastapi/issues/2510

**Comment by jurasofish** [comment]
Thanks!

**Comment by gjcarneiro** [comment]
> This is a duplicate of #2510

IMHO it is not a dup of that bug, it's merely related.

* #2510 is: "field with nullable=True doesn't reflect that in the openapi schema/docs;
* this bug is "can we please infer that nullable=True if the field python type is Optional[something]"

We just had a bunch of lost time in my team due to this bug:
* `myfield Optional[int]` correctly causes the field to be marked as optional in the docs
* but I changed to `parent_group_id: Optional[int] = Field(...)` in order to ensure that the field is always present, even if null; but now the docs don't reflect that it is optional.  Unfortunately I missed that the docs changed this way.

As a result, a bunch of time wasted back and forth between frontend and backend devs, because the documentation didn't clearly reflect which fields were optional.

I shame, really.  FastAPI is amazing at time saving and automatically generating docs.  Would be so much nicer if it automatically deduced this one as well.

**Comment by jlplenio** [comment]
> * but I changed to `parent_group_id: Optional[int] = Field(...)` in order to ensure that the field is always present, even if null; but now the docs don't reflect that it is optional.  Unfortunately I missed that the docs changed this way.

With this setup, did you add the nullable=True flag, to indicated in the docs, that the required field is nullable?

**Comment by gjcarneiro** [comment]
Yeah, from reading these comments, I figured adding `nullable=True` would solve my problem.  And it did.  

I just wish things were a bit more automatic.  IMHO, with `Optional[int] = Field(...)`, the Optional should imply that the field is nullable and required, by default.  Instead, by default it implies that the field is required and not nullable.

**Comment by fzyzcjy** [comment]
Hi, how can I see `nullable` in swagger/openapi using the following code:

```
class A(BaseModel):
    a: Optional[int]
```

Thanks!

Or, do I have to write down `Field(nullable=True)` every time? IMHO fastapi says it does not repeat itself, so writing `Optional` and `nullable=True` seems to be a duplication.

**Comment by amacfie** [comment]
@jurasofish please reopen

**Comment by ivantichy** [comment]
> Hi, how can I see `nullable` in swagger/openapi using the following code:
> 
> ```
> class A(BaseModel):
>     a: Optional[int]
> ```
> 
> Thanks!
> 
> Or, do I have to write down `Field(nullable=True)` every time? IMHO fastapi says it does not repeat itself, so writing `Optional` and `nullable=True` seems to be a duplication.

IMHO "required" and "nullable" are two different things.
https://stackoverflow.com/questions/45575493/what-does-required-in-openapi-really-mean

**Comment by cbartz** [comment]
> This is arguably a bug. But the root issue is in Pydantic, see [pydantic/pydantic#1270](https://github.com/pydantic/pydantic/issues/1270).

As this is clearly a bug, this should be stated somewhere in the FastAPI docs (with the workaround given by specifying ``nullable=True`` for each optional field).

I've used auto-generated clients that refused to work with a FastAPI server because null values were sent for optional but non-nullable fields.
